stripe/src/main/java/com/stripe/android/view/CardInputView.java

```diff
@@ -40,9 +55,120 @@ private void initView() {
+            @Override
+            public void onFocusChange(View v, boolean hasFocus) {
+                if (hasFocus) {
+                    scrollLeft();
+                }
+            }
+        });
+
+        mExpiryDateEditText.setOnFocusChangeListener(new OnFocusChangeListener() {
+            @Override
+            public void onFocusChange(View v, boolean hasFocus) {
+                if (hasFocus) {
+                    scrollRight();
+                }
+            }
+        });
+
+        mCvcNumberEditText.setOnFocusChangeListener(new OnFocusChangeListener() {
+            @Override
+            public void onFocusChange(View v, boolean hasFocus) {
+                if (hasFocus) {
+                    scrollRight();
+                }
+            }
+        });
+
+        mExpiryDateEditText.setOnKeyListener(new OnKeyListener() {
+            @Override
+            public boolean onKey(View v, int keyCode, KeyEvent event) {
+                if (keyCode == KeyEvent.KEYCODE_DEL
+                        && event.getAction() == KeyEvent.ACTION_DOWN) {
+                    if (mExpiryDateEditText.getText().length() == 0) {
+                        mCardNumberEditText.requestFocus();
+                    }
+                }
+                return false;
+            }
+        });
+
+        mCardNumberEditText.setCardNumberCompleteListener(
+                new CardNumberEditText.CardNumberCompleteListener() {
+                    @Override
+                    public void onCardNumberComplete() {
+                        scrollRight();
+                    }
+                });
+
+        mCardNumberEditText.requestFocus();
+    }
+
+    @Override
+    protected void onAttachedToWindow() {
+        super.onAttachedToWindow();
+
+    }
+
+    @Override
+    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
+        super.onLayout(changed, left, top, right, bottom);
+        // We set this length the first time the control is laid out, because the
+        // value for the actual width of mScrollView prior to this will be zero.
+        if (!mInitializedFlag) {
+            mInitializedFlag = true;
+            mScrollViewWidth = mScrollView.getMeasuredWidth();
+            mScrollToPostion = mScrollViewWidth;
+            ViewGroup.LayoutParams widthParams = mCardNumberSpace.getLayoutParams();
+            widthParams.width = mScrollViewWidth;
+            mCardNumberSpace.setLayoutParams(widthParams);
+        }
+    }
+
+    private void scrollRight() {
+        if (!mCardNumberIsViewed) {
+            return;
+        }
+        mCardNumberIsViewed = false;
+        updateScrollToPosition();
+        mScrollView.smoothScrollBy(mScrollToPostion, 0);
+        mExpiryDateEditText.setVisibility(View.VISIBLE);
+        mCvcNumberEditText.setVisibility(View.VISIBLE);
+        mExpiryDateEditText.requestFocus();
+    }
+
+    private void scrollLeft() {
```

Comment:
I like this pattern of updating UI from centralized places. Any reason, ooc, that we wouldn't create one `updateUI` method that everyone calls. It would make sure it knows what state the UI is in based on the state of the individual fields and would be one centralized place for handling all UI updates for that view.

Again, just a curious thought. Not blocking the PR.